pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -90,34 +123,94 @@ public String segmentStoppedConsuming(
+      LOGGER.info(responseStr);
+      return responseStr;
+    } catch (Throwable t) {
+      LOGGER.error("Exception handling request", t);
+      throw new WebApplicationException(t);
+    }
   }
 
-  /*
   @POST
-  @Path(SegmentCompletionProtocol.MSG_TYPE_STOPPED_CONSUMING)
+  @Path(SegmentCompletionProtocol.MSG_TYPE_COMMIT)
+  @Consumes(MediaType.MULTIPART_FORM_DATA)
   @Produces(MediaType.APPLICATION_JSON)
   public String segmentCommit(
       @QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE_ID) String instanceId,
       @QueryParam(SegmentCompletionProtocol.PARAM_SEGMENT_NAME) String segmentName,
-      @QueryParam(SegmentCompletionProtocol.PARAM_OFFSET) long offset
-  ) {
-
+      @QueryParam(SegmentCompletionProtocol.PARAM_OFFSET) long offset,
+      FormDataMultiPart multiPart
+  )
+  {
     SegmentCompletionProtocol.Request.Params requestParams = new SegmentCompletionProtocol.Request.Params();
     requestParams.withInstanceId(instanceId).withSegmentName(segmentName).withOffset(offset);
     LOGGER.info(requestParams.toString());
 
-    SegmentCompletionProtocol.Response response = SegmentCompletionManager.getInstance().segmentStoppedConsuming(
-        requestParams);
+    final SegmentCompletionManager segmentCompletionManager = SegmentCompletionManager.getInstance();
+    SegmentCompletionProtocol.Response response = segmentCompletionManager.segmentCommitStart(requestParams, false);
+    if (response.equals(SegmentCompletionProtocol.RESP_COMMIT_CONTINUE)) {
+      // Get the segment and put it in the right place.
+      boolean success = uploadSegment(multiPart, instanceId, segmentName);
+
+      response = segmentCompletionManager.segmentCommitEnd(requestParams, success, false);
+    }
+
+    LOGGER.info("Response: instance={}  segment={} status={} offset={}", requestParams.getInstanceId(), requestParams.getSegmentName(),
+        response.getStatus(), response.getOffset());
+
+    return response.toJsonString();
+  }
+
+
+  private boolean uploadSegment(FormDataMultiPart multiPart, final String instanceId, final String segmentName) {
+    FileUploadPathProvider provider = new FileUploadPathProvider(_controllerConf);
+
+    Map<String, List<FormDataBodyPart>> map = multiPart.getFields();
+    if (!validateMultiPart(map, segmentName)) {
+      return false;
+    }
+    final String name = map.keySet().iterator().next();
+    final FormDataBodyPart bodyPart = map.get(name).get(0);
+    InputStream is = bodyPart.getValueAs(InputStream.class);
+
+    File tmpFile = new File(provider.getTmpDir(), name);
+
+    try {
+      provider.saveStreamAs(is, tmpFile);
+      LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+      final String rawTableName = llcSegmentName.getTableName();
+      final File tableDir = new File(provider.getBaseDataDir(), rawTableName);
+      final File segmentFile = new File(tableDir, segmentName);
+
+      synchronized (_helixResourceManager) {
```

Comment:
I have added comments on the scenarios in which this can happen. I have changed the synchronization to be on SegmentCompletionManager.